File: ios/TeeTimeCaddie/TeeTimeCaddie/Source/analytics/FirebaseEventPlugin.swift
//
//  FirebaseEventPlugin.swift
//  TeeTimeCaddie
//
//  Created by Brad Ball on 5/21/23.
//

import Foundation
import TeeTimeCaddieKit
import FirebaseAnalytics

class FirebaseEventPlugin: EventPlugin {
    func logEvent(event: AnalyticsEvent) -> Bool {
        Analytics.logEvent(event.name, parameters: event.asMap)
        return true
    }
    
    func logScreenView(screen: AnalyticsScreen, displayMethod: ScreenType) -> Bool {

        
        var params: [String:Any] = screen.parameters
        
        params[AnalyticsParameterScreenName] = screen.name
        params[AnalyticsParameterScreenClass] = displayMethod.displayName
        
        Analytics.logEvent(AnalyticsEventScreenView, parameters: params)

        return true
    }
    
    func setUserId(userId: String) {
        Analytics.setUserID(userId)
    }
}

File: android/app/src/main/java/net/bradball/teetimecaddie/android/analytics/FirebaseEventPlugin.kt
package net.bradball.teetimecaddie.android.analytics

import androidx.core.os.bundleOf
import com.google.firebase.Firebase
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.analytics
import com.google.firebase.analytics.logEvent
import net.bradball.teetimecaddie.core.analytics.AnalyticsEvent
import net.bradball.teetimecaddie.core.analytics.AnalyticsScreen
import net.bradball.teetimecaddie.core.analytics.EventPlugin
import net.bradball.teetimecaddie.core.analytics.ScreenType

class FirebaseEventPlugin: EventPlugin {
    override fun logEvent(event: AnalyticsEvent): Boolean {

        val params = event.asMap.toList().toTypedArray()

        Firebase.analytics.logEvent(event.name, bundleOf(*params))
        return true
    }

    override fun setUserId(userId: String) {
        Firebase.analytics.setUserId(userId)
    }

    override fun logScreenView(screen: AnalyticsScreen, displayMethod: ScreenType): Boolean {
        Firebase.analytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
            param(FirebaseAnalytics.Param.SCREEN_NAME, screen.viewName)
            param(FirebaseAnalytics.Param.SCREEN_CLASS, screen.name)
            screen.parameters.forEach { (key, value) -> param(key, value) }
            param("displayMethod", displayMethod.displayName)
        }

        return true
    }

}